### PR TITLE
Chore/upgrade ng and bs

### PR DESCRIPTION
**Pull request by jnizet**

### CODE REVIEW

File: frontend/src/app/form-control-validation.directive.ts [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
-  ngControl: NgControl;
+  constructor(@Optional() private ngControl: NgControl) { }
 
-  constructor() { }
-
-  @HostBinding('class.has-danger') get hasDanger() {
-    return this.ngControl && this.ngControl.dirty && this.ngControl.invalid;
+  @HostBinding('class.is-invalid') get isInvalid() {
+    return this.ngControl
+      && this.ngControl
```
Comment (cexbrayat) [inline frontend/src/app/form-control-validation.directive.ts]: This line is duplicated

File: frontend/src/app/login/login.component.html [hunk 2]
```diff
@@ -11,13 +11,13 @@ <h1>Connexion</h1>
       <div class="form-group">
         <label class="from-control-label" for="login">Identifiant</label>
         <input id="login" class="form-control" name="login" [(ngModel)]="credentials.login" required #loginCtrl="ngModel">
-        <div class="from-control-feedback" *ngIf="loginCtrl.control.dirty && loginCtrl.control.hasError('required')">L'identifiant est obligatoire</div>
+        <div class="invalid-feedback" *ngIf="loginCtrl.hasError('required')">L'identifiant est obligatoire</div>
```
[the remark this answers is not in the document]
Comment (cexbrayat) [reply frontend/src/app/login/login.component.html]: Nevermind, I guess it's `is-invalid`, which changes when the field is touched, that reveals the message.
These kind of directives are powerful but sometimes it's hard to follow 😅

File: frontend/src/app/app.module.ts [hunk 3]
```diff
@@ -87,6 +87,11 @@ import { ChargeTypeResolverService } from './charge-type-resolver.service';
 import { ChargeService } from './charge.service';
 import { ChargesResolverService } from './charges-resolver.service';
 import { PersonChargeEditComponent } from './person-charge-edit/person-charge-edit.component';
+import { registerLocaleData } from '@angular/common';
+import localeFr from '@angular/common/locales/fr';
+import { EmptyResponseBodyErrorInterceptorService } from './empty-response-body-error-interceptor.service';
+
+registerLocaleData(localeFr);
```
[the remark this answers is not in the document]
Comment (jnizet) [reply frontend/src/app/app.module.ts]: Yes, since we only have one language supported, and since we already customize the locale by hand, I thought it would be easier and faster to leave the build simple, and to hard-code everything to French.

File: frontend/src/app/charge-types/charge-types.component.html [hunk 4]
```diff
@@ -5,7 +5,7 @@ <h1>Types de charge</h1>
   <div class="col-sm text-muted">{{ chargeType.category.name }}</div>
   <div class="col-sm text-muted">
     <ng-container *ngIf="chargeType.maxMonthlyAmount">
-      max. {{ chargeType.maxMonthlyAmount | currency:'EUR':true }} / mois
+      max. {{ chargeType.maxMonthlyAmount | currency:'EUR':'symbol-narrow' }} / mois
```
[the remark this answers is not in the document]
Comment (jnizet) [reply frontend/src/app/charge-types/charge-types.component.html]: Ah, nice. Didn't know it was the default.

File: frontend/src/app/login/login.component.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (jnizet) [reply frontend/src/app/login/login.component.html]: It's just bootstrap helping us: `.invalid-feedback` has a display of none unless its sibling form control has `.is-invalid`. So it's pure CSS, but it's helpful because it allows avoiding the repetition of the complex condition (touched or enclosing form submitted).

**Comment by jnizet** [comment]
Thanks Cedric. As you've seen, I have tried improving feedback on forms. Even if, in this case, mandatory fields are kind of obvious, I find it more user-friendly to let the user try to submit the form and display all errors after the submission (or after he touched a field). That way, he/she doesn't have to wonder why the submit button is disabled and why no error message is displayed, only because he/she never dirtied the required fields.

**Comment by cexbrayat** [comment]
Yes I like it this way :+1: Ideally we should have some kind of visual indication that the field is required, but that does the job.

**Comment by acrepet** [comment]
I have briefly traveled across the app. LGTM!
